### DISCUSSION TITLE
debugging a python lambda function code

### DISCUSSION THREAD

**Discussion by Ore-stack** [tofetpuzo/Kaisen#34] [General]
```
import json 
import boto3
def lambda_handler(event, context):
    s3 = boto3.client("s3")
    sns = boto3.client('sns')
    print(event)
    bucket = event['Records'][0]['s3']['bucket']['name']
    key = event['Records'][0]['s3']['object']['key']
    print(bucket)
    print(key)
    data = s3.get_object(Bucket=bucket, Key=key)
    print(data)
    contents = data['Body'].read()
    total_words = contents.split()
    print('The word count in the file ' ,key, ' is ' , len(total_words))
    snsArn = 'asdagfagdfdaExercise'
    message = "The word count in the file "+key+" is " + str(len(total_words))
    response = sns.publish(TopicArn = snArn,Message = message ,Subject='Word Count Result')
```

**Comment by tofetpuzo** [comment]
Run this code let see what happens?

```
import json
import boto3
def lambda_handler(event, context):
    s3 = boto3.client("s3")
    sns = boto3.client('sns')
    print(event)
```

**Reply by Ore-stack** [reply]
Okay, on it

**Reply by tofetpuzo** [reply]
what was your result?

**Reply by Ore-stack** [reply]
Response
{
  "errorMessage": "Syntax error in module 'lambda_function': expected an indented block after function definition on line 3 (lambda_function.py, line 4)",
  "errorType": "Runtime.UserCodeSyntaxError",
  "requestId": "",
  "stackTrace": [
    "  File \"/var/task/lambda_function.py\" Line 4\n    s3 = boto3.client(\"s3\")\n"
  ]
}

**Reply by tofetpuzo** [reply]
you need to format the code ...

**Reply by tofetpuzo** [reply]
I have modified it?

**Reply by Ore-stack** [reply]
format what code please?

**Reply by tofetpuzo** [reply]
just code the code again , I have changed it. the code was not indented properly

**Reply by Ore-stack** [reply]
Oh ok

**Reply by Ore-stack** [reply]
Test Event Name
(unsaved) test event

Response
null

**Reply by tofetpuzo** [reply]
It means you don't have an `event or context` , you will need to define them

**Reply by tofetpuzo** [reply]
```
def lambda_handler(event, context):
    # Access event data
    print("Received event:", event)

    # Access context data
    print("Function name:", context.function_name)
    print("Memory limit (MB):", context.memory_limit_in_mb)
    print("Remaining time (ms):", context.get_remaining_time_in_millis())

    # Your function logic
    return {
        'statusCode': 200,
        'body': 'Hello from Lambda!'
    }

```

**Reply by Ore-stack** [reply]
I'll try this now

**Reply by tofetpuzo** [reply]
> Test Event Name (unsaved) test event
> 
> Response null

so you can see why you are getting `key error` event doesn't have anything and you want to access it using this code 
```
bucket = event['Records'][0]['s3']['bucket']['name']
```
 event is what you receive back from aws lambda func

**Reply by Ore-stack** [reply]
oh ok

**Reply by tofetpuzo** [reply]
In AWS Lambda, an event is the data that triggers the execution of your Lambda function.

**Reply by tofetpuzo** [reply]
this is how an event should look like, if it is null then aws doesn't acknowledge you have anything inside s3 bucket
```
{
    "Records": [
        {
            "s3": {
                "bucket": {
                    "name": "name-of-bucket"
                },
                "object": {
                    "key": "file.txt"
                }
            }
        }
    ]
}
```

**Reply by Ore-stack** [reply]
I got this when I tried the code

Test Event Name
(unsaved) test event

Response
{
  "statusCode": 200,
  "body": "Hello from Lambda!"
}

**Reply by Ore-stack** [reply]
this is how an event should look like, if it is null then aws doesn't acknowledge you have anything inside s3 bucket

>  ```
>  {
>  "Records": [ {            
>                 "s3": {
>                     "bucket": {
>                      "name": "name-of-bucket"
>                  },
>                  "object": {
>                     "key": "file.txt" 
>                 } 
>              }
>          }
>     ]  }
>  ```
> 

An event is written in json format I guess?

**Reply by tofetpuzo** [reply]
yeah it is in a json format.

**Reply by Ore-stack** [reply]
> > this is how an event should look like, if it is null then aws doesn't acknowledge you have anything inside s3 bucket
> > ```
> > 
> > {
> > 
> >     "Records": [
> > 
> >         {
> > 
> >             "s3": {
> > 
> >                 "bucket": {
> > 
> >                     "name": "name-of-bucket"
> > 
> >                 },
> > 
> >                 "object": {
> > 
> >                     "key": "file.txt"
> > 
> >                 }
> > 
> >             }
> > 
> >         }
> > 
> >     ]
> > 
> > }
> > ```
> 
> An event is written in json format I guess?

Right! Can I just copy this and paste in my Event JSON?

**Reply by tofetpuzo** [reply]
No you have to fill in all the things you want like `file.txt` has to be the document you want to read from or find the `len` of text in it  from. `"name-of-bucket"`  it has to be the actual name of your s3 bucket. 

Have a read here [aws_lambda](https://docs.aws.amazon.com/lambda/latest/dg/python-handler.html) and let me know what you  don't understand

**Reply by Ore-stack** [reply]
Yeah, I know I have to replace those two things

**Reply by Ore-stack** [reply]
Event has been created successfully, so, what's next?

**Reply by tofetpuzo** [reply]
You don't have to pass it to the event , you should already have data from your s3 bucket. run your initial code let us see what happens.

**Reply by Ore-stack** [reply]
The one I sent earlier which contains my sns arn?

**Reply by Ore-stack** [reply]
> You don't have to pass it to the event , you should already have data from your s3 bucket. run your initial code let us see what happens.

Response
{
  "errorMessage": "An error occurred (InvalidParameter) when calling the Publish operation: Invalid parameter: TopicArn",
  "errorType": "InvalidParameterException",
  "requestId": "0b319336-8cec-48df-9744-f4c983919ee1",
  "stackTrace": [
    "  File \"/var/task/lambda_function.py\", line 18, in lambda_handler\n    response = sns.publish(TopicArn = snsArn, Message = message, Subject ='Word Count Result')\n",
    "  File \"/var/runtime/botocore/client.py\", line 565, in _api_call\n    return self._make_api_call(operation_name, kwargs)\n",
    "  File \"/var/runtime/botocore/client.py\", line 1021, in _make_api_call\n    raise error_class(parsed_response, operation_name)\n"
  ]
}

**Reply by tofetpuzo** [reply]
ok let us start from the beginning let me see if you correctly configured your s3 bucket


```
import json
import boto3

def lambda_handler(event, context):
    s3 = boto3.client('s3')

    bucket_name = 'your-bucket-name'
    object_key = 'your-object-key'

    try:
        response = s3.get_object(Bucket=bucket_name, Key=object_key)
        content = response['Body'].read().decode('utf-8')
        print(content)
        
        return {
            'statusCode': 200,
            'body': json.dumps('Success!'),
            'content': content
        }

    except Exception as e:
        print(e)
        return {
            'statusCode': 500,
            'body': json.dumps(f'Error: {str(e)}')
        }
```

**Reply by Ore-stack** [reply]
> ok let us start from the beginning let me see if you correctly configured your s3 bucket
> 
> 
> 
> 
> 
> ```
> 
> import json
> 
> import boto3
> 
> 
> 
> def lambda_handler(event, context):
> 
>     s3 = boto3.client('s3')
> 
> 
> 
>     bucket_name = 'your-bucket-name'
> 
>     object_key = 'your-object-key'
> 
> 
> 
>     try:
> 
>         response = s3.get_object(Bucket=bucket_name, Key=object_key)
> 
>         content = response['Body'].read().decode('utf-8')
> 
>         print(content)
> 
>         
> 
>         return {
> 
>             'statusCode': 200,
> 
>             'body': json.dumps('Success!'),
> 
>             'content': content
> 
>         }
> 
> 
> 
>     except Exception as e:
> 
>         print(e)
> 
>         return {
> 
>             'statusCode': 500,
> 
>             'body': json.dumps(f'Error: {str(e)}')
> 
>         }
> 
> ```

How do you want to see this?

**Reply by Ore-stack** [reply]
or should I say what exactly do you want to see?

**Comment by Ore-stack** [comment]
Okay


ok let us start from the beginning let me see if you correctly configured your s3 bucket


```
import json
      import boto3
      
      def lambda_handler(event, context):
          s3 = boto3.client('s3')
      
          bucket_name = 'your-bucket-name'
          object_key = 'your-object-key'
      
          try:
              response = s3.get_object(Bucket=bucket_name, Key=object_key)
              content = response['Body'].read().decode('utf-8')
              print(content)
              
              return {
                  'statusCode': 200,
                  'body': json.dumps('Success!'),
                  'content': content
              }
      
          except Exception as e:
              print(e)
              return {
                  'statusCode': 500,
                  'body': json.dumps(f'Error: {str(e)}')
              }
```

**Reply by tofetpuzo** [reply]
do you get any response?

**Reply by Ore-stack** [reply]
Yes, I did but in the console only, I guess I didn't configure the SNS Topic quite well

**Reply by tofetpuzo** [reply]
so you did spot the error? great work. so run the code let's see what happens.

**Reply by Ore-stack** [reply]
Not yet cause if I did, I'll get the response in my subscribed email address

**Reply by tofetpuzo** [reply]
did you look at your snsArn?

**Reply by Ore-stack** [reply]
I'm yet to do that

**Reply by tofetpuzo** [reply]
try and do that you have 3 of these `:::` it should be 1 per command.

**Reply by Ore-stack** [reply]
Okay I'll

**Reply by tofetpuzo** [reply]
how are you getting on?

**Comment by Ore-stack** [comment]
On Tuesday, September 3, 2024, 9:28 PM, temis ***@***.***> wrote:

> how are you getting on?
 Work has not allowed me get on o

**Reply by tofetpuzo** [reply]
Have you tried the solution I provided?

**Reply by Ore-stack** [reply]
I've been busy with work and classes but I'll do that and get back to you on Friday unfailingly if that's alright

**Reply by Ore-stack** [reply]
This is the solution given to the lambda function python code

import json
import boto3
def lambda_handler(event, context):
   s3 = boto3.client("s3")
   sns = boto3.client('sns')
   print(event)
   bucket = event['Records'][0]['s3']['bucket']['name']
   key = event['Records'][0]['s3']['object']['key']
   print(bucket)
   print(key)
   data = s3.get_object(Bucket=bucket, Key=key)
   print(data)
   contents = data['Body'].read()
   total_words = contents.split()
   print('The word count in the file ' ,key, ' is ' , len(total_words))
   
   snsArn = 'paste the sns arn'
   message = "The word count in the file "+key+" is  " + str(len(total_words))
   response = sns.publish(TopicArn = snsArn,Message = message ,Subject='Word Count Result')

**Reply by Ore-stack** [reply]
It's finally worked

![image](https://github.com/user-attachments/assets/85b75f9a-5209-4ed8-a45b-2a102b8f41a0)

**Reply by tofetpuzo** [reply]
yeah it does work 😉

**Comment by ohnogaurav** [comment]
Try this mate and let me know:  

```python
import json
import boto3

def lambda_handler(event, context):
    s3 = boto3.client("s3")
    sns = boto3.client('sns')
    print(event)
    
    bucket = event['Records'][0]['s3']['bucket']['name']
    key = event['Records'][0]['s3']['object']['key']
    print(bucket)
    print(key)
    
    data = s3.get_object(Bucket=bucket, Key=key)
    print(data)
    
    contents = data['Body'].read().decode('utf-8')
    total_words = contents.split()
    print('The word count in the file', key, 'is', len(total_words))
    
    snsArn = 'arn:aws:sns:region:account-id:topic-name'  # Update with your actual SNS ARN
    message = "The word count in the file " + key + " is " + str(len(total_words))
    
    response = sns.publish(
        TopicArn=snsArn,
        Message=message,
        Subject='Word Count Result'
    )
    return {
        'statusCode': 200,
        'body': json.dumps('SNS notification sent successfully')
    }

**Reply by tofetpuzo** [reply]
this looks good
